third-party/realdds/src/dds-device-broadcaster.cpp

```diff
@@ -6,233 +6,154 @@
-                    _new_reader_joined = true;
-                    _owner->_trigger_msg_send = true;
+                    eprosima::fastdds::dds::ConditionSeq active_conditions;
+                    wait_set.wait( active_conditions, eprosima::fastrtps::c_TimeInfinite );
+                    if( _stopped.get_trigger_value() )
+                        break;
+                    // Let multiple broadcasts gather and do it only once
+                    std::this_thread::sleep_for( std::chrono::milliseconds( 50 ) );
+                    _ready_for_broadcast.set_trigger_value( false );
+
+                    std::lock_guard< std::mutex > lock( _broadcasters_mutex );
+                    LOG_DEBUG( _writer->topic()->get_participant()->name() << ": broadcasting" );
+                    for( dds_device_broadcaster const * broadcaster : _broadcasters )
+                        broadcaster->broadcast();
                 }
-                _owner->_new_client_cv.notify_all();
+                LOG_DEBUG( _writer->topic()->get_participant()->name() << ": broadcaster thread stopped" );
             } );
-        }
-        else if( info.current_count_change == -1 )
-        {
-        }
-        else
-        {
-            LOG_ERROR( std::to_string( info.current_count_change )
-                       << " is not a valid value for on_publication_matched" );
-        }
     }
 
-private:
-    dds_device_broadcaster * _owner;
-};
-
-
-dds_device_broadcaster::dds_device_broadcaster( std::shared_ptr< dds_participant > const & participant )
-    : _trigger_msg_send( false )
-    , _participant( participant )
-    , _publisher( std::make_shared< dds_publisher >( participant ) )
-    , _dds_device_dispatcher( 10 )
-    , _new_client_handler( [this]( dispatcher::cancellable_timer ) {
-        // We wait until the new reader callback indicate a new reader has joined or until the active object is stopped
-        if( _active )
+    ~broadcast_manager()
+    {
+        if( _th.joinable() )
         {
-            std::unique_lock< std::mutex > lock( _new_client_mutex );
-            _new_client_cv.wait( lock, [this]() { return ! _active || _trigger_msg_send.load(); } );
-            if( _active && _trigger_msg_send.load() )
-            {
-                _trigger_msg_send = false;
-                _dds_device_dispatcher.invoke( [this]( dispatcher::cancellable_timer ) {
-                    for( auto const & sn_and_handle : _device_handle_by_sn )
-                    {
-                        if( sn_and_handle.second.client_listener->_new_reader_joined )
-                        {
-                            if( send_device_info_msg( sn_and_handle.second ) )
-                            {
-                                sn_and_handle.second.client_listener->_new_reader_joined = false;
-                            }
-                        }
-                    }
-                } );
-            }
+            _stopped.set_trigger_value( true );
```

Comment:
No need to set for non joinable threads?

Reply:
If it's not joinable then it's not running and nothing to wait on